src/lib.rs

```diff
@@ -405,6 +482,29 @@ impl TrustAnchor<'_> {
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct CertificateRevocationListDer<'a>(Der<'a>);
 
+#[cfg(feature = "alloc")]
+impl DecodePem for Vec<CertificateRevocationListDer<'static>> {
```

Comment:
It seems surprising/weird that `DecodePem` is implemented for `PrivatePkcs8KeyDer` but not for `CertificateDer` (and implemented for `Vec<CertificateDer>` but not `Vec<PrivatePkcs8KeyDer>`).

It seems to me that the fundamental required trait method here is more like `from_item(item: pem::Item) -> Option<Self>`.

I think we can rebuild the other things on top of that -- it should compose well with `FromIterator` and things like `Iterator::filter()`.

Reply:
Yes I agree it currently feels odd, with an implicit assumption that some types are plural (certs, crls) and others singular (keys). Will experiment some more.

Reply:
Thinking about this more, I feel like `Item` is not quite the right core abstraction -- most people really aren't interested in what it does, which is abstracting over types that can appear in PEM objects. It feels like the core abstraction is that a PEM stream yields `(SectionType, Vec<u8>)` items and then the core pki-types could implement something like this:

```rust
trait FromPem: From<Vec<u8>> {
     const KIND: SectionType;
}
```

Reply:
I have been trying this, but it comes a bit unstuck when it comes to `impl FromPem for PrivateKeyDer`, as that has to recognise three possible section types. However I think the previous suggestion still is an improvement -- there's a sketch of this in the commit I just pushed.

Reply:
Ah yeah, that's a tricky edge case. Maybe the associated const is a slice?

Reply:
OK, a little more movement on this. Unfortunately I have abandoned the trait entirely, as it would require rustc 1.75 to be able to return iterators in a trait function. That's a shame. Now there is a macro instead which provides the common constructor/iterator functions.